cpp/include/cuspatial/experimental/detail/linestring_intersection_with_duplicates.cuh

```diff
@@ -0,0 +1,297 @@
+                     range(lhs_segment_ids.begin(), lhs_segment_ids.end()),
+                     range(rhs_linestring_ids.begin(), rhs_linestring_ids.end()),
+                     range(rhs_segment_ids.begin(), rhs_segment_ids.end())};
+  }
+
+  /// Return the number of pairs in the intermediates
+  auto size() { return offsets.size() - 1; }
+};
+
+/**
+ * @brief Kernel to compute intersection and store the result and look-back indices to outputs.
+ *
+ * Each thread operates on one segment in `multilinestrings1`, iterates over all segments in the
+ * other multilinestring.
+ */
+template <typename MultiLinestringRange1,
+          typename MultiLinestringRange2,
+          typename TempIt1,
+          typename TempIt2,
+          typename Offsets1,
+          typename Offsets2,
+          typename Offsets3,
+          typename IdRanges,
+          typename OutputIt1,
+          typename OutputIt2>
+void __global__ pairwise_linestring_intersection_simple(MultiLinestringRange1 multilinestrings1,
+                                                        MultiLinestringRange2 multilinestrings2,
+                                                        TempIt1 n_points_stored,
+                                                        TempIt2 n_segments_stored,
+                                                        Offsets1 num_points_offsets_first,
+                                                        Offsets2 num_segments_offsets_first,
+                                                        Offsets3 num_points_per_pair_first,
+                                                        IdRanges point_ids_range,
+                                                        IdRanges segment_ids_range,
+                                                        OutputIt1 points_first,
+                                                        OutputIt2 segments_first)
+{
+  using T       = typename MultiLinestringRange1::element_t;
+  using types_t = uint8_t;
+  using count_t = iterator_value_type<Offsets1>;
+  for (auto idx = threadIdx.x + blockIdx.x * blockDim.x; idx < multilinestrings1.num_points();
+       idx += gridDim.x * blockDim.x) {
+    auto const part_idx = multilinestrings1.part_idx_from_point_idx(idx);
+    if (!multilinestrings1.is_valid_segment_id(idx, part_idx)) continue;
+    auto const lhs_linestring_idx = multilinestrings1.intra_part_idx(part_idx);
+    auto const lhs_segment_idx    = multilinestrings1.intra_point_idx(idx);
+    auto [a, b]                   = multilinestrings1.segment(idx);
+    auto const geometry_idx       = multilinestrings1.geometry_idx_from_part_idx(part_idx);
+    auto const multilinestring2   = multilinestrings2[geometry_idx];
+    auto const geometry_collection_offset =
+      num_points_offsets_first[geometry_idx] + num_segments_offsets_first[geometry_idx];
+
+    for (auto rhs_linestring_idx = 0; rhs_linestring_idx < multilinestring2.size();
+         ++rhs_linestring_idx) {
```

Comment:
Now that you have a `range` type, could be?

```suggestion
    for (auto rhs_linestring_idx : cuspatial::range(0, multilinestring2.size() - 1) {
```

Reply:
Ooh, looks nice

Reply:
Hang on... You can't do that. `Range` is composed of iterators, not raw values. You wouldn't want to dereference `0` as an address.

I think what you are looking for is `range(counting_iterator(0), counting_iterator(multilinestring2.size()-1))`, but I'm not sure which is more succinct. Also, in c++ 20, this is what we need: https://en.cppreference.com/w/cpp/ranges/iota_view

Reply:
Actually, you are looking for https://github.com/harrism/ranger :)